src/Uno.UI/Microsoft/UI/Xaml/Controls/TreeView/TreeViewNode.cs

```diff
@@ -0,0 +1,255 @@
+				}
+				m_itemsSource = value;
+				m_itemsDataSource = value != null ? new ItemsSourceView(value) : null;
+				if (m_itemsDataSource != null)
+				{
+					m_itemsDataSource.CollectionChanged += OnItemsSourceChanged;
+				}
+				SyncChildrenNodesWithItemsSource();
+			}
+		}
+
+		private void OnItemsSourceChanged(object sender, NotifyCollectionChangedEventArgs args)
+		{
+			switch (args.Action)
+			{
+				case NotifyCollectionChangedAction.Add:
+					{
+						// TreeViewNode and ItemsSource will update each other when data changes.
+						// For ItemsSource -> TreeViewNode changes, m_itemsDataSource.Count() > Children().Size()
+						// We'll add the new node to children collection.
+						// For TreeViewNode -> ItemsSource changes, m_itemsDataSource.Count() == Children().Size()
+						// the node is already in children collection, we don't want to update TreeViewNode again here.
+						if (m_itemsDataSource.Count != Children.Count)
+						{
+							AddToChildrenNodes(args.NewStartingIndex, args.NewItems.Count);
+						}
+						break;
+					}
+
+				case NotifyCollectionChangedAction.Remove:
+					{
+						// TreeViewNode and ItemsSource will update each other when data changes.
+						// For ItemsSource -> TreeViewNode changes, m_itemsDataSource.Count() < Children().Size()
+						// We'll remove the node from children collection.
+						// For TreeViewNode -> ItemsSource changes, m_itemsDataSource.Count() == Children().Size()
+						// the node is already removed, we don't want to update TreeViewNode again here.
+						if (m_itemsDataSource.Count != Children.Count)
+						{
+							RemoveFromChildrenNodes(args.OldStartingIndex, args.OldItems.Count);
+						}
+						break;
+					}
+
+				case NotifyCollectionChangedAction.Reset:
+					{
+						SyncChildrenNodesWithItemsSource();
+						break;
+					}
+
+				case NotifyCollectionChangedAction.Replace:
+					{
+						RemoveFromChildrenNodes(args.OldStartingIndex, args.OldItems.Count);
+						AddToChildrenNodes(args.NewStartingIndex, args.NewItems.Count);
+						break;
+					}
+			}
+		}
+
+		private void AddToChildrenNodes(int index, int count)
+		{
+			for (int i = index + count - 1; i >= index; i--)
+			{
+				var item = m_itemsDataSource.GetAt(i);
+				var node = new TreeViewNode();
+				node.Content = item;
+				((TreeViewNodeVector)Children).InsertAt(index, node, false /* updateItemsSource */);
+			}
+		}
+
+		private void RemoveFromChildrenNodes(int index, int count)
+		{
+			for (int i = 0; i < count; i++)
+			{
+				((TreeViewNodeVector)Children).RemoveAt(index, false /* updateItemsSource */);
+			}
+		}
+
+		void SyncChildrenNodesWithItemsSource()
+		{
+			if (!AreChildrenNodesEqualToItemsSource())
+			{
+				var children = (TreeViewNodeVector)Children;
+				children.Clear(false /* updateItemsSource */);
+
+				var size = m_itemsDataSource != null ? m_itemsDataSource.Count : 0;
+				for (var i = 0; i < size; i++)
+				{
+					var item = m_itemsDataSource.GetAt(i);
```

Comment:
![Codacy](https://app.codacy.com/assets/images/favicon.png) Issue found: ['m_itemsDataSource' is null on at least one execution path.](https://app.codacy.com/gh/unoplatform/uno/pullRequest?prid=5156819)